3.0/m_xlinetools.cpp

```diff
@@ -0,0 +1,602 @@
+
+<helpop key="xcount" value="/XCOUNT -type=<X-line type|*> -mask=[!]<m> -reason=[!]<r> -source=[!]<s> -set=[-]<t> -duration=[-+]<t> -expires=[+]<t>
+
+Returns a count of matching X-lines of the specified type (or all types). Mask(m) supports CIDR, reason(r) can include spaces, source(s)
+is a nick, server, or <Config>. 'Prefix your value for these 3 arguments with a '!' to negate the match. 't' is a
+time-string (seconds or 1y2w3d4h5m6s) and can be prefixed with '+' or '-' to adjust matching. All arguments are optional.">
+
+<helpop key="xremove" value="/XREMOVE -type=<X-line type|*> -mask=[!]<m> -reason=[!]<r> -source=[!]<s> -set=[-]<t> -duration=[-+]<t> -expires=[+]<t>
+
+Removes matching X-lines of the specified type (or all types). Mask(m) supports CIDR, reason(r) can include spaces, source(s)
+is a nick, server, or <Config>. 'Prefix your value for these 3 arguments with a '!' to negate the match. 't' is a
+time-string (seconds or 1y2w3d4h5m6s) and can be prefixed with '+' or '-' to adjust matching. All arguments are optional.
+Use /XSEARCH to test before removing.">
+
+<helpop key="xsearch" value="/XSEARCH -type=<X-line type|*> -mask=[!]<m> -reason=[!]<r> -source=[!]<s> -set=[-]<t> -duration=[-+]<t> -expires=[+]<t>
+
+Lists matching X-lines of the specified type (or all types). Mask(m) supports CIDR, reason(r) can include spaces, source(s)
+is a nick, server, or <Config>. Prefix your value for these 3 arguments with a '!' to negate the match. 't' is a
+time-string (seconds or 1y2w3d4h5m6s) and can be prefixed with '+' or '-' to adjust matching. All arguments are optional.">
+
+ */
+
+
+#include "inspircd.h"
+#include "xline.h"
+
+struct Criteria
+{
+	std::string type;
+	std::string mask;
+	std::string reason;
+	std::string source;
+	std::string set;
+	std::string duration;
+	std::string expires;
+
+	Criteria() { }
+	Criteria(const std::string& t, const std::string& m, const std::string& r, const std::string& s)
+		: type(t)
+		, mask(m)
+		, reason(r)
+		, source(s)
+	{
+	}
+};
+
+namespace
+{
+	bool HasPermission(LocalUser* user, std::string type)
+	{
+		if (type.length() <= 2)
+			type.append("LINE");
+
+		return user->HasPermission(type);
+	}
+
+	bool ProcessArgs(const CommandBase::Params& params, Criteria& args)
+	{
+		if (!params.size())
+			return false;
+
+		// Track arg reason to include space separated params
+		bool argreason = false;
+
+		const std::string mtype("-type=");
+		const std::string mmask("-mask=");
+		const std::string mreason("-reason=");
+		const std::string msource("-source=");
+		const std::string msetby("-setby=");
+		const std::string mset("-set=");
+		const std::string mduration("-duration=");
+		const std::string mexpires("-expires=");
+
+		for (std::vector<std::string>::const_iterator p = params.begin(); p != params.end(); ++p)
+		{
+			const std::string param = *p;
+
+			if (param.find(mtype) != std::string::npos)
+			{
+				argreason = false;
+				const std::string& val(param.substr(mtype.length()));
```

Comment:
I'm not sure if this was in the original but this (and below) should be a `std::string` not a `std::string&`.